### PR TITLE
many: build drivers tree when current mount is not the target mount

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
In some cases (when using the snapd install API or when installing
from initramfs), the place where the kernel snap / components used for the
installation are mounted is different to the final location in the
installed system. This change considers this so the drivers tree is
generated with symlinks pointing to the final expected location.

Additionally, fix `muinstaller-{core,real}` tests for core24.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
This includes https://github.com/snapcore/snapd/pull/14041 - I've split as maybe we want that before the rest of this PR.

File: kernel/kernel_drivers.go [hunk 1]
```diff
@@ -217,6 +253,21 @@ type KernelDriversTreeOptions struct {
 	KernelInstall bool
 }
 
+// MountPoints describes mount points for a snap or a component.
+type MountPoints struct {
+	// CurrentMntPt is where the container to be installed is currently
+	// available
+	CurrentMntPt string
```
Comment (pedronis) [inline kernel/kernel_drivers.go:260]: not sure we need the MntPt bit given the name of the struct?

File: kernel/kernel_drivers.go [hunk 2]
```diff
@@ -92,20 +91,26 @@ func createFirmwareSymlinks(fwMount, fwDest string) error {
 		return err
 	}
 
+	fwTarget := filepath.Join(fwMount.TargetMntPt, "firmware")
```
Comment (pedronis) [inline kernel/kernel_drivers.go:94]: idea: maybe MountPoints should have CurrentPath and TargetPath methods that do filepath.Join?

File: kernel/kernel_drivers.go [hunk 3]
```diff
@@ -148,41 +153,58 @@ func createModulesSubtree(kernelMount, kernelTree, kversion string, kmodsConts [
...
+	return nil
+}
+
+func createKernelModulesSymlinks(modsRoot, kMntPt string) error {
 	for _, d := range []string{"kernel", "vdso"} {
 		lname := filepath.Join(modsRoot, d)
-		to := filepath.Join(earlyMntDir, d)
+		to := filepath.Join(kMntPt, d)
+		// We might be re-creating, first remove
+		os.Remove(lname)
```
Comment (pedronis) [inline kernel/kernel_drivers.go:183]: do we need to do Remove before other Symlink calls ?

**Review by pedronis — COMMENTED** [review]
did a first quick pass on kernel_drivers.go

File: kernel/kernel_drivers.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply kernel/kernel_drivers.go]: Done

File: kernel/kernel_drivers.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply kernel/kernel_drivers.go:183]: This is the name of the link that we are creating, so I don't think there could be a collision, unless I am missing something here.

**Comment by alfonsosanchezbeato** [comment]
@pedronis thanks, I've addressed your comments now

File: tests/lib/tools/setup_nested_hybrid_system.sh [hunk 4]
```diff
@@ -155,7 +158,11 @@ run_muinstaller() {
     image_path="${NESTED_IMAGES_DIR}/${image_name}"
     kpartx -asv "${image_path}"
     fatlabel /dev/disk/by-label/ubuntu-seed UBUNTU-SEED
-    kpartx -d "${image_path}"
+    if ! kpartx -d "${image_path}"; then
+        # Some times there are random failures, let's wait and re-try
```
Comment (bboozzoo) [inline tests/lib/tools/setup_nested_hybrid_system.sh:162]:
```suggestion
        # Sometimes there are random failures, let's wait and re-try
```

File: overlord/snapstate/backend/setup.go [hunk 5]
```diff
@@ -348,22 +352,32 @@ func mergeCompSideInfosUpdatingRev(comps1, comps2 []*snap.ComponentSideInfo) (me
 // finalComps, for the kernel/revision specified by ksnapName/ksnapRev.
 func moveKModsComponentsState(currentComps, finalComps []*snap.ComponentSideInfo, ksnapName string, ksnapRev snap.Revision, cleanErrMsg string) (err error) {
 	cpi := snap.MinimalSnapContainerPlaceInfo(ksnapName, ksnapRev)
+	kMntPts := kernel.MountPoints{
+		Current: cpi.MountDir(),
+		Target:  cpi.MountDir()}
```
Comment (bboozzoo) [inline overlord/snapstate/backend/setup.go:357]:
```suggestion
		Target:  cpi.MountDir()
	}
```

File: kernel/kernel_drivers.go [hunk 6]
```diff
@@ -195,6 +217,20 @@ func setupModsFromComp(kernelTree, kversion string, kmodsConts []snap.ContainerP
...
 
+	// Change symlinks to target ones when needed
+	for _, mci := range comps {
+		if mci.Mounts.CurrentEqualsTarget() {
+			continue
+		}
+		lname := filepath.Join(compsRoot, mci.Name)
+		to := mci.Mounts.UnderTargetPath("modules", kversion)
+		// remove old link
+		os.Remove(lname)
```
Comment (bboozzoo) [inline kernel/kernel_drivers.go:228]: do we to check the error?
```suggestion
		if err := os.Remove(lname); err != nil && !errors.Is(err, fs.ErrNotExist) { 
			return fmt.Errorf("cannot remove old link: %w", err)
		}	
```

File: gadget/install/content.go [hunk 7]
```diff
@@ -139,7 +139,13 @@ func writeFilesystemContent(laidOut *gadget.LaidOutStructure, kSnapInfo *KernelS
 		destDir := kernel.DriversTreeDir(destRoot, kSnapInfo.Name, kSnapInfo.Revision)
 		logger.Noticef("building drivers tree in %s", destDir)
 
-		if err := kernelEnsureKernelDriversTree(kSnapInfo.MountPoint, destDir, nil,
+		kTargetSysMntPt := snap.MinimalSnapContainerPlaceInfo(kSnapInfo.Name, kSnapInfo.Revision)
+		if err := kernelEnsureKernelDriversTree(
+			kernel.MountPoints{
+				Current: kSnapInfo.MountPoint,
+				Target:  kTargetSysMntPt.MountDir()},
```
Comment (bboozzoo) [inline gadget/install/content.go:146]:
```suggestion
				Target:  kTargetSysMntPt.MountDir()
			},
```

File: gadget/install/content_test.go [hunk 8]
```diff
@@ -290,8 +290,8 @@ func (s *contentTestSuite) testWriteFilesystemContentDriversTree(c *C, isCore bo
 		IsCore:           isCore,
 	}
 
-	restore = install.MockKernelEnsureKernelDriversTree(func(kSnapRoot, destDir string, kmodsConts []snap.ContainerPlaceInfo, opts *kernel.KernelDriversTreeOptions) (err error) {
-		c.Check(kSnapRoot, Equals, kMntPoint)
+	restore = install.MockKernelEnsureKernelDriversTree(func(kMntPts kernel.MountPoints, comps []kernel.ModulesCompInfo, destDir string, opts *kernel.KernelDriversTreeOptions) (err error) {
+		c.Check(kMntPts.Target, Equals, kMntPoint)
```
Comment (pedronis) [inline gadget/install/content_test.go:294]: should we check Current?

File: kernel/kernel_drivers.go [hunk 9]
```diff
@@ -92,20 +91,26 @@ func createFirmwareSymlinks(fwMount, fwDest string) error {
...
 	for _, node := range entries {
-		lpath := filepath.Join(fwDest, node.Name())
-		origPath := filepath.Join(fwOrig, node.Name())
 		switch node.Type() {
 		case 0, fs.ModeDir:
+			// "updates" is included in (latest) kernel snaps but
+			// is empty, and we use if for firmware shipped in
+			// components, so we ignore it.
+			if node.Name() == "updates" {
+				continue
```
Comment (pedronis) [inline kernel/kernel_drivers.go:102]: this bit doesn't seem tested

File: kernel/kernel_drivers.go (hunk 6 above)
Comment (pedronis) [reply kernel/kernel_drivers.go:228]: the osSymlink itself will fail if the remove is relevant and fails

File: kernel/kernel_drivers.go [hunk 10]
```diff
@@ -217,6 +253,33 @@ type KernelDriversTreeOptions struct {
...
+func (mp *MountPoints) UnderTargetPath(dirs ...string) string {
+	return filepath.Join(append([]string{mp.Target}, dirs...)...)
+}
+
+func (mp *MountPoints) CurrentEqualsTarget() bool {
+	return mp.Current == mp.Target
+}
+
+// ModulesCompInfo contains mount points for a component plus its name.
+type ModulesCompInfo struct {
```
Comment (pedronis) [inline kernel/kernel_drivers.go:278]: my preference would be to also call this ModulesCompMountPoints ?  there's an another Info in this package and is very different, so having both be called Info is confusing to me

**Review by pedronis — COMMENTED** [review]
thanks for the changes, did another pass

File: gadget/install/content.go (hunk 7 above)
Comment (alfonsosanchezbeato) [reply gadget/install/content.go:146]: Changed

File: gadget/install/content_test.go (hunk 8 above)
Comment (alfonsosanchezbeato) [reply gadget/install/content_test.go:294]: Done

File: kernel/kernel_drivers.go (hunk 9 above)
Comment (alfonsosanchezbeato) [reply kernel/kernel_drivers.go:102]: Test added

File: kernel/kernel_drivers.go (hunk 10 above)
Comment (alfonsosanchezbeato) [reply kernel/kernel_drivers.go:278]: Changed

File: overlord/snapstate/backend/setup.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/setup.go:357]: Changed

File: tests/lib/tools/setup_nested_hybrid_system.sh (hunk 4 above)
Comment (alfonsosanchezbeato) [reply tests/lib/tools/setup_nested_hybrid_system.sh:162]: Changed

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis @bboozzoo thanks for the reviews, I've addressed your comments now. I also have added a change to the testing infrastructure now that I know more about a failure related to lxd installation.

File: gadget/install/content_test.go [hunk 11]
```diff
@@ -290,8 +290,9 @@ func (s *contentTestSuite) testWriteFilesystemContentDriversTree(c *C, isCore bo
 		IsCore:           isCore,
 	}
 
-	restore = install.MockKernelEnsureKernelDriversTree(func(kSnapRoot, destDir string, kmodsConts []snap.ContainerPlaceInfo, opts *kernel.KernelDriversTreeOptions) (err error) {
-		c.Check(kSnapRoot, Equals, kMntPoint)
+	restore = install.MockKernelEnsureKernelDriversTree(func(kMntPts kernel.MountPoints, compsMntPts []kernel.ModulesCompMountPoints, destDir string, opts *kernel.KernelDriversTreeOptions) (err error) {
+		c.Check(kMntPts, Equals,
+			kernel.MountPoints{Current: kMntPoint, Target: kMntPoint})
```
Comment (pedronis) [inline gadget/install/content_test.go:295]: do we need a test where these end up being different? looking at the code that we changed in content.go is not clear they will always be the same?

Comment (alfonsosanchezbeato) [reply gadget/install/content_test.go]: I've added a test for this now

**Review by bboozzoo — APPROVED** [review]
LGTM

**Comment by bboozzoo** [comment]
@alfonsosanchezbeato please rebase on top of master

**Pull request merged**